csrc/rsa_cipher.cpp

```diff
@@ -29,56 +54,80 @@ JNIEXPORT jint JNICALL Java_com_amazon_corretto_crypto_provider_RsaCipher_cipher
     try {
         raii_env env(pEnv);
 
-        EvpKeyContext *ctx = reinterpret_cast<EvpKeyContext *>(ctxHandle);
-
-        RSA* r = EVP_PKEY_get0_RSA(ctx->getKey()); // Doesn't need to be freed
-
-        if (!r) {
-            throw_java_ex(EX_NPE, "Null RSA key");
-        }
         if (!input) {
             throw_java_ex(EX_NPE, "Null input array");
         }
         if (!output) {
             throw_java_ex(EX_NPE, "Null output array");
         }
 
+        EvpKeyContext *ctx = reinterpret_cast<EvpKeyContext *>(ctxHandle);
+
+        EVP_PKEY_CTX *keyCtx = ctx->getKeyCtx();
+        if (keyCtx == nullptr) {
+            keyCtx = ctx->setKeyCtx(EVP_PKEY_CTX_new(ctx->getKey(), /*engine*/nullptr));
+        }
```

Comment:
Reading through the EvpKeyFactory and NativeResource code, it looks like the RSA key is ["generated"](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/EvpKeyFactory.java#L53-L64) by [converting](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/csrc/java_evp_keys.cpp#L128) from DER, which [allocates](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/csrc/keyutils.cpp#L25) the `EVP_PKEY` struct. the `*EVP_PKEY` is then returned to EvpKey, which takes ownership of it as a NativeResource. Doesn't each RsaCipher instance already [have its own (java) EvpKey](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/RsaCipher.java#L79), that encapsulates a (native) `*EVP_PKEY` pointer as a (java) NativeResource?

is your concern here that multiple threads would use the same EvpKey instance with multiple ciphers and differing OAEP params such that the threads trample on each other?

> It might well be worth changing EvpKey so that instead of being backed by an EvpKeyContext it is backed directly by an EVP_PKEY.

you're referring to [this EvpKeyContext](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/csrc/keyutils.h#L18)? it looks like it's backed by [either/both of a `*EVP_PKEY` or/and a `*EVP_PKEY_CTX`](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/csrc/keyutils.h#L90-L91)

Reply:
An [EvpKey](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/EvpKey.java)'s NativeResource is backed by an `EvpKeyContext` (rather than directly by an `EVP_PKEY`. This can be seen throughout [java_evp_keys.cpp](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/csrc/java_evp_keys.cpp).

This same `EvpKeyContext` is what is passed down as the pointer by [RsaCipher](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/RsaCipher.java#L162-L163) and received (in your code) in [rsa_cipher.cpp](https://github.com/WillChilds-Klein/amazon-corretto-crypto-provider/blob/pss-params/csrc/rsa_cipher.cpp#L63). While the `EvpKeyContext` does contain an `EVP_PKEY` (from `EvpKey`) and that object *is* thread-safe (and more importantly not mutated), you are now also [adding an EVP_PKEY_CTX](https://github.com/WillChilds-Klein/amazon-corretto-crypto-provider/blob/pss-params/csrc/rsa_cipher.cpp#L65-L68) to the `EvpKeyContext` object. `EVP_PKEY_CTX` is mutated (and thus, not thread-safe). This means that multiple instances of `RsaCipher` may be using the same `EvpKey` and thus the same `EvpKeyContext` and thus the same `EVP_PKEY` (**safe**) and the same `EVP_PKEY_CTX` (**not safe**) simultaneously.

Reply:
ah, i see. thanks for the explanation, that distinction makes a lot of sense.. i think my confusion was twofold -- 1. confusing terms (namely native `EvpKeyContext` with AWS-LC `EVP_PKEY_CTX`) and 2. not realizing that `EVP_PKEY` is r/o and threadsafe. OpenSSL's [docs](https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_CTX_new.html) also make `EVP_PKEY_CTX`'s unsafety clear:

> The `EVP_PKEY_CTX` structure is an opaque public key algorithm context used by the OpenSSL high-level public key API. Contexts **MUST NOT** be shared between threads: that is it is not permissible to use the same context simultaneously in two threads.

regarding your first comment:

> It might well be worth changing `EvpKey` so that instead of being backed by an `EvpKeyContext `it is backed directly by an `EVP_PKEY`. This would be good evidence that I made a mistake in the design.

what's the thinking here? i get under this scheme the NativeResource would now point to an `EVP_PKEY` instead of our wrapper `EvpKeyContext`. so we'd rip out `EvpKeyContext` entirely and just keep a NativeResource pointing to the `EVP_PKEY_CTX` on all of RsaCipher, RsaGen, and EvpSignature instances?

Reply:
Basically.

However, this change can be a fast follow. I'll approve this and we'll get my PR fixing things merged tomorrow if possible.